bulletroboy/exoforce/exoforce_hardware.py

```diff
@@ -0,0 +1,266 @@
+import numpy as np
+from pyquaternion import Quaternion
+
+from .exoforce import ExoForce
+from ..utils.utils import Topics, Services
+
+from roboy_simulation_msgs.msg import Collision
+from roboy_middleware_msgs.msg import MotorCommand, MotorState
+from roboy_middleware_msgs.srv import ControlMode
+from geometry_msgs.msg import PoseStamped
+
+from rclpy.callback_groups import ReentrantCallbackGroup
+
+FAST_PWM_TENSION = 300
+MIN_PWM_TENSION = 150
+YES_OPTIONS = ["Yes","yes","y",""]
+
+POS_CONTROL_THRESHOLD = 0.1
+
+class ExoforceHW(ExoForce):
+	def __init__(self, cage_conf):
+		"""
+		Args:
+			cage_conf (CageConfiguration): Cage configuration defined in the configuration file.
+		
+		"""
+		super().__init__(cage_conf, "exoforce", POS_CONTROL_THRESHOLD)
+		self.callback_group = ReentrantCallbackGroup()
+
+		self.create_subscription(Collision, Topics.MAPPED_COLLISIONS, self.collision_listener, 1)
+		self.create_subscription(Collision, 'roboy/simulation/roboy/collision_hw', self.collision_listener, 1)
+		# self.create_subscription(PoseStamped, Topics.OP_EF_POSES, self.ef_pos_listener, 1)
+		self.create_subscription(PoseStamped, Topics.VR_HEADSET_POSES, self.operator_ef_pos_listener, 10, callback_group=self.callback_group)
+		
+		self.roboy_plexus_ready = False
+		choice = input("Connect to roboy plexus? (Yes/no): ")
+		if choice in YES_OPTIONS:
+			self.create_subscription(MotorState, Topics.MOTOR_STATE, self.motor_state_listener, 1, callback_group=self.callback_group)
+			self.motor_command_publisher = self.create_publisher(MotorCommand, Topics.MOTOR_COMMAND, 1)
+
+			self.control_mode_client = self.create_client(ControlMode, Services.CONTROL_MODE)
+			while not self.control_mode_client.wait_for_service(timeout_sec=1.0):
+				self.get_logger().info('roboy_plexus not available, waiting again...')
+			self.get_logger().info('Succesfull connection to roboy plexus!')
```

Comment:
not guaranteed

Reply:
see https://github.com/Roboy/bulletroboy/pull/25#issuecomment-760128816